APIServer Distributed Tracing Alpha (#28843)
kubernetes/website: content/en/docs/concepts/cluster-administration/system-traces.md

---
title: Traces For Kubernetes System Components
reviewers:
- logicalhan
- lilic
content_type: concept
weight: 60
---

<!-- overview -->

> 

Reviewer: This should use a [feature state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state) to show that it's not a generally available feature.

Author: done

System component traces record the latency of and relationships between operations in the cluster.

> Kubernetes components emit traces in [OTLP format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification) using the grpc exporter, and can be collected and routed to tracing backends using the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector#-opentelemetry-collector).

Reviewer 2: ```suggestion
Kubernetes components emit traces in [open telemetry format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification) using the gRPC exporter and can be collected and routed to tracing backends using the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector#-opentelemetry-collector).
```

Reviewer: “OTLP” seems Ok to me.

Reviewer 2: It's abbreviated from [`OpenTelemetry Protocol`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md)  and I personally prefer using long names in documentation, especially since it's a relatively new concept.

Author: Changed to OpenTelemetry Protocol

<!-- body -->

## Stability

Tracing instrumentation is still under active development, and may change in a variety of ways. This includes span names, attached attributes, instrumented endpoints, etc. There are currently no guarantees of backwards-compatibility for tracing instrumentation.

## Trace Collection

For a complete guide to collecting traces and using the collector, see [Getting Started with the OpenTelemetry Collector](https://opentelemetry.io/docs/collector/getting-started/). However, there are a few things to note that are specific to Kubernetes components.

> By default, kubernetes componnets export traces using the grpc exporter for OTLP on the [IANA OpenTelemetry port](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=opentelemetry), 4317. As an example, if the collector is running as a sidecar to a kuberntes component, the following receiver configuration will collect spans and log them to stdout:

Reviewer: ```suggestion
By default, Kubernetes components export traces using the grpc exporter for OTLP on the [IANA OpenTelemetry port](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=opentelemetry), 4317. As an example, if the collector is running as a sidecar to a Kubernetes component, the following receiver configuration will collect spans and log them to standard output:
```

```yaml
receivers:
  otlp:
    protocols:
      grpc:
exporters:
  # Replace this exporter with the exporter for your backend
  logging:
    logLevel: debug
service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [logging]
```

## Component traces

### kube-apiserver traces

> The kube-apiserver generates spans for incoming http requests, and for outgoing requests to webhooks, etcd, and re-entrant requests.  It propagates the [W3C Trace Context](https://www.w3.org/TR/trace-context/) with outgoing requests, but does not make use of the trace context attached to incoming requests, as the kube-apiserver is often a public endpoint.

Reviewer 2: ```suggestion
The kube-apiserver generates spans for incoming HTTP requests, and for outgoing requests to webhooks, etcd, and re-entrant requests.  It propagates the [W3C Trace Context](https://www.w3.org/TR/trace-context/) with outgoing requests but does not make use of the trace context attached to incoming requests, as the kube-apiserver is often a public endpoint.
```

#### Enabling tracing in the kube-apiserver

> To enable tracing, enable the [`APIServerTracing` feature gate](/docs/reference/command-line-tools-reference/feature-gates/) on the kube-apiserver.  Also, provide the kube-apiserver with a tracing configration file with `--tracing-config-file=<path-to-config>`.  This is an example config that records spans for 1 in 10000 requests, and uses the default OpenTelemetry endpoint:

Reviewer: nit:
```suggestion
To enable tracing, enable the `APIServerTracing` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) on the kube-apiserver.  Also, provide the kube-apiserver with a tracing configration file with `--tracing-config-file=<path-to-config>`.  This is an example config that records spans for 1 in 10000 requests, and uses the default OpenTelemetry endpoint:
```

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: TracingConfiguration

> samplingRatePerMillion: 100

Reviewer 2: Since the other option we provide to configure is the endpoint, we should either mention it or just override it here as a way of self-documenting. What do you think?

Reviewer: Could be a commented-out line for the endpoint?

Reviewer 2: Yes I think that sounds good as well, just more of a hint that this is the other thing you can override and how to do it.

Author: Done

---
title: Traces For Kubernetes System Components
reviewers:
- logicalhan
- lilic
content_type: concept
weight: 60
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

System component traces record the latency of and relationships between operations in the cluster.

Kubernetes components emit traces using the [OpenTelemetry Protocol](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification) with the gRPC exporter and can be collected and routed to tracing backends using the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector#-opentelemetry-collector).

<!-- body -->

## Stability

> Tracing instrumentation is still under active development, and may change in a variety of ways. This includes span names, attached attributes, instrumented endpoints, etc. There are currently no guarantees of backwards-compatibility for tracing instrumentation.

Reviewer: I'd move this to end end of the page.
Also, maybe change:
> There are currently no guarantees of backwards-compatibility for tracing instrumentation.

to
> Until this feature graduates to stable, there are  no guarantees of backwards compatibility
> for tracing instrumentation.

Author: Done

```yaml
receivers:
  otlp:
    protocols:
      grpc:
exporters:
  # Replace this exporter with the exporter for your backend
  logging:
    logLevel: debug
service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [logging]
```

## Component traces

### kube-apiserver traces

The kube-apiserver generates spans for incoming HTTP requests, and for outgoing requests to webhooks, etcd, and re-entrant requests.  It propagates the [W3C Trace Context](https://www.w3.org/TR/trace-context/) with outgoing requests but does not make use of the trace context attached to incoming requests, as the kube-apiserver is often a public endpoint.

#### Enabling tracing in the kube-apiserver

To enable tracing, enable the `APIServerTracing` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) on the kube-apiserver.  Also, provide the kube-apiserver with a tracing configration file with `--tracing-config-file=<path-to-config>`.  This is an example config that records spans for 1 in 10000 requests, and uses the default OpenTelemetry endpoint:

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: TracingConfiguration

> # endpoint: localhost:4317 is the default

Reviewer: ```suggestion
# default value
#endpoint: localhost:4317
```

---
title: Traces For Kubernetes System Components
reviewers:
- logicalhan
- lilic
content_type: concept
weight: 60
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

System component traces record the latency of and relationships between operations in the cluster.

> Kubernetes components emit traces using the [OpenTelemetry Protocol](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification) with the gRPC exporter and can be collected and routed to tracing backends using the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector#-opentelemetry-collector).

Reviewer 3: s/the/an/ [OpenTelemetry Collector]

Author: done
